airflow/migrations/versions/e655c0453f75_add_taskmap_and_map_id_on_taskinstance.py

```diff
@@ -0,0 +1,120 @@
+# or more contributor license agreements.  See the NOTICE file
+# distributed with this work for additional information
+# regarding copyright ownership.  The ASF licenses this file
+# to you under the Apache License, Version 2.0 (the
+# "License"); you may not use this file except in compliance
+# with the License.  You may obtain a copy of the License at
+#
+#   http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing,
+# software distributed under the License is distributed on an
+# "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+# KIND, either express or implied.  See the License for the
+# specific language governing permissions and limitations
+# under the License.
+
+"""Add TaskMap and map_index on TaskInstance.
+
+Revision ID: e655c0453f75
+Revises: 587bdf053233
+Create Date: 2021-12-13 22:59:41.052584
+"""
+
+from alembic import op
+from sqlalchemy import Column, ForeignKeyConstraint, Integer
+
+from airflow.models.base import StringID
+from airflow.utils.sqlalchemy import ExtendedJSON
+
+# Revision identifiers, used by Alembic.
+revision = "e655c0453f75"
+down_revision = "587bdf053233"
+branch_labels = None
+depends_on = None
+
+
+def upgrade():
+    """Add TaskMap and map_index on TaskInstance."""
+    # We need to first remove constraints on task_reschedule since they depend on task_instance.
+    with op.batch_alter_table("task_reschedule") as batch_op:
+        batch_op.drop_constraint("task_reschedule_ti_fkey", "foreignkey")
+        batch_op.drop_index("idx_task_reschedule_dag_task_run")
+
+    # Change task_instance's primary key.
+    with op.batch_alter_table("task_instance") as batch_op:
+        # I think we always use this name for TaskInstance after 7b2661a43ba3?
+        batch_op.drop_constraint("task_instance_pkey", type_="primary")
+        batch_op.add_column(Column("map_index", Integer, nullable=False, default=-1))
+        batch_op.create_primary_key("task_instance_pkey", ["dag_id", "task_id", "run_id", "map_index"])
+
+    # Re-create task_reschedule's constraints.
+    with op.batch_alter_table("task_reschedule") as batch_op:
+        batch_op.add_column(Column("map_index", Integer, nullable=False, default=-1))
+        batch_op.create_foreign_key(
+            "task_reschedule_ti_fkey",
+            "task_instance",
+            ["dag_id", "task_id", "run_id", "map_index"],
+            ["dag_id", "task_id", "run_id", "map_index"],
+            ondelete="CASCADE",
+        )
+        batch_op.create_index(
+            "idx_task_reschedule_dag_task_run",
+            ["dag_id", "task_id", "run_id", "map_index"],
+            unique=False,
+        )
+
+    # Create task_map.
+    op.create_table(
+        "task_map",
+        Column("dag_id", StringID(), primary_key=True),
+        Column("task_id", StringID(), primary_key=True),
+        Column("run_id", StringID(), primary_key=True),
+        Column("map_index", Integer, primary_key=True),
```

Comment:
Not that it matters, but for consistency should we have the same default here as we do on TI's map_index column?

Reply:
Actually I think this _shouldn’t_ be `-1` because `TI.map_index == -1` means the TI is _not_ mapped. Let me check if SQLAlchemy has an unsigned interger column type…

Reply:
Oh TIL the SQL standards do not ever define an unsigned integer type!

Reply:
Added a CheckConstraint and a check in `__init__` for this instead.

Reply:
The default on `TI.map_index` is mainly added so we don’t need to change too much existing code (otherwise we’d need to add a ton of `map_index=-1`), but this being a new class, I think I prefer being explicit instead.